HHS/Head-Start-TTADP
# [TTAHUB-2560] dedupe ARG and create timeseries functions, restore missing ARGFRs

## Description of change

The first migration creates and runs a stored procedure for merging duplicate ARGs. It can and probably should be rerun in the future, but for now it's just here so that when we add missing Activity Report Goal Field Responses, there's only one target Activity Report Goal.

The second migration creates a stored procedure that accepts any audited table name (as a string literal) and uses the audit log to create an expanded version of that table containing one timeseries record for each historical state of each record in the audited table. Each record has the data as it was in that version, plus two additional columns `timeband_start` and `timeband_end` that contain the timestamps for the start and end of the band of time that the original record was in that state

The utility of this is demonstrated by the simplicity of the final query, where we join the activity report to the goal responses as they looked at the time the report was approved just by adding `ar."approvedAt" BETWEEN timeband_start AND timeband_end`

## How to test

There's a lot of different things to check here. One thing to try is making updates to your local database and then using `SELECT create_timeseries_from_audit_log('TableYouUpdated');` to generate `TableYouUpdated_timeseries` and then inspecting to see that it represented your updates faithfully.

also try running
```
SELECT DISTINCT
  arg."activityReportId" ar_id,
  arg."goalId" goal_id,
  arfr.response root_cause_on_ar,
  gfr.response root_cause_on_goal,
  ar."calculatedStatus" ar_status
FROM "Goals" g
JOIN "ActivityReportGoals" arg
  ON g.id = arg."goalId"
JOIN "ActivityReports" ar
  ON arg."activityReportId" = ar.id
LEFT JOIN "ActivityReportGoalFieldResponses" arfr
  ON arg.id = arfr."activityReportGoalId"
LEFT JOIN "GoalFieldResponses" gfr
  ON g.id = gfr."goalId"
WHERE g."goalTemplateId" = 19017
  AND ar."createdAt" BETWEEN '2023-04-28' AND '2024-02-01' ;
```
which in the original Jira shows null `root_cause_on_ar` values where `root_cause_on_goal` is not null. There should be no more of those in that time period or at any other.

## Issue(s)

* https://jira.acf.gov/browse/TTAHUB-2560
* https://jira.acf.gov/browse/TTAHUB-2789


## Checklists

### Every PR



### Before merge to main


### Production Deploy


### After merge/deploy
